src/orange/OrangeParams.cc

```diff
@@ -95,6 +95,82 @@ OrangeInput input_from_json(std::string filename)
     return result;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the maximum num of universe levels within the geometry
+ *
+ * A single-universe geometry will have a max_depth of 1.
+ * The operator() is a recursive function, so for external callers, uid should
+ * be the root universe id.
+ */
+struct MaxDepthCalculator
+{
+    HostVal<OrangeParamsData> const& data_;
+    std::map<UniverseId, size_type> depths;
+
+    // Construct from host data
+    MaxDepthCalculator(HostVal<OrangeParamsData> const& data) : data_(data) {}
+
+    // Calculate max depth
+    size_type operator()(UniverseId const& uid)
+    {
+        CELER_EXPECT(uid);
+
+        // Check for cached value
+        if (depths.find(uid) != depths.end())
+        {
+            return depths[uid];
+        }
```

Comment:
```suggestion
Best-practices note: doing `find` + `[]` is repeating the same binary lookup; best practice is to do:
```c++
        auto&& [iter, inserted] = depths.insert({uid, {}});
        if (!inserted)
        {
            // Return cached value
            return iter->second;
        }
```

and then at the end of the function, set `iter->second` to the value you calculated. This way you only have one lookup per function instead of two.

And if you're not caching the value, you still shouldn't use `[]` after `find`: just save `if (auto iter = mymap.find(foo); iter != mymap.end()) { return iter->second; }`